### DISCUSSION TITLE
When using Pull Request model how to safely delete local branch

### DISCUSSION THREAD

**Discussion by plastikfan** [community/community#49327] [Pull Requests]
### Select Topic Area

Question

### Body

I am migrating to the Pull Request model and am worried about how to safely delete the local branch after the pull request has been rebased(with squash) back onto the default branch.

After doing the rebase merge in the browser, I am still left with my local branch. Since these changes have been pushed and rebased, there should be no changes in my local branch that are missing from the upstream default branch.

So when I deleted my local branch, I was surprised to find this happen:

```
 plastikfan@Janus  ~/dev/github/go/snivilised/lucretia λ   master  git branch -d feat/modify-age-prompt
warning: deleting branch 'feat/modify-age-prompt' that has been merged to
         'refs/remotes/origin/feat/modify-age-prompt', but not yet merged to HEAD.
```

I don't know what this message means, but if we look at the Git Graph:

![after-rebase-git-graph](https://user-images.githubusercontent.com/49785914/223154730-8d3a7eba-8586-4783-aea8-bb89c4d1f9f0.jpg)

We can see the the commit ___feat(greet): modify age prompt (#9)___ (#169f557f) looks ok, it has a single ancestor now, as opposed the the other --no-ff merges which have 2 ancestors.

So the upstream has all the code, with nothing missing. So what does git mean when it issues the warning:

```
warning: deleting branch 'feat/modify-age-prompt' that has been merged to
         'refs/remotes/origin/feat/modify-age-prompt', but not yet merged to HEAD
```

When using the pull request model, what is the correct course of action to delete the local branch without fear of doing the 'wrong' thing. I say this, because I want to avoid any chance of getting the repo into a state where we have a detached HEAD, which is in itself, confusing and worrying.

PS, the repository I am wroking with is my own, so I don't need to fork it

Cheers.

**Comment by plastikfan** [comment]
Oops, this is now going from bad to worse.

I was trying to test that the rebase/squash method on github was doing what I expected (I wanted to see that the rebase was creating a single commit from the multiple commits (just 2) in the feature branch), so this is what I did:

- start new feature branch __feat/add-refreshments__
- make first commit ___add basic refreshement(66aca848)___
- push commit 66aca848
- make second commit ___add treats(0727808a)___
- push commit 0727808a
- on github, rebase __feat/add-refreshments__ onto __master__ => ci completed ok

at this point I thought I was now done (apart from deleting local branch).

So:

* first I detached my local branch from the remote: __git branch --unset-upstream__
* Then I deleted the local branch as I have done before: __git branch -d feat/add-refreshments__

```
 plastikfan@Janus  ~/dev/github/go/snivilised/lucretia λ   master  git branch -d feat/add-refreshments
error: The branch 'feat/add-refreshments' is not fully merged.
If you are sure you want to delete it, run 'git branch -D feat/add-refreshments'.
```

which did not make any sense to me.

On my local master branch, I performed a pull, which retrieved remote (squashed rebase) content as expected:

```
 plastikfan@Janus  ~/dev/github/go/snivilised/lucretia λ   master  git pull origin master
remote: Enumerating objects: 12, done.
remote: Counting objects: 100% (12/12), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 10 (delta 5), reused 5 (delta 3), pack-reused 0
Unpacking objects: 100% (10/10), 1.48 KiB | 760.00 KiB/s, done.
From https://github.com/snivilised/lucretia
 * branch            master     -> FETCH_HEAD
   81e4606..0727808  master     -> origin/master
Updating 81e4606..0727808
Fast-forward
 greeter.go          |  2 +-
 refreshment_test.go | 19 +++++++++++++++++++
 refreshments.go     | 21 +++++++++++++++++++++
 3 files changed, 41 insertions(+), 1 deletion(-)
 create mode 100644 refreshment_test.go
 create mode 100644 refreshments.go
 plastikfan@Janus  ~/dev/github/go/snivilised/lucretia
```

The changes indeed are present in my local master and so nothing is missing,but when I look at the graph:

![fake-lost-commits-squash-rebase](https://user-images.githubusercontent.com/49785914/223176515-f80a2848-4d95-4828-a240-d113ce8f9d18.jpg)

which now looks a complete and utter mess.

So what's going wrong here? What is the correct workflow for the Pull Request model? In particluar how to get rid of the local feature branch which should now not contain anything that is NOT present upstream.

**Comment by spenserblack** [comment]
If you squash-merge, then you are *not* merging your branch's commits into `master`, but creating a *new* commit containing all of the changes of your PR.

Also, if you rebase-merge, you are creating new commits with the same diffs and messages as the PR's commits, *but different SHAs.* This is because rebase-merged commits can have different parent commits than the commits in the PR.

Git is warning you about deleting your branch because those *are* different commits than what is in `master`, even if the file contents of `master` and your other branch are the same.

This is a bit of an informal method, but what I do is simply run `git fetch --prune`, and delete all branches that were reported as deleted by the remote. This assumes that, if I'm deleting a branch remotely, I'm OK with that branch getting deleted locally, too (e.g. after a PR is merged). Additionally, there are several CLI tools that will do this for you, if you do some searching.

**Reply by plastikfan** [reply]
thanks for responding, I just need to take a moment to consider your advice.

**Reply by plastikfan** [reply]
Actually, the reason why I want to squash commits is because my next task is to activate the linear history rule which ensures that all commits on the main branch have only a single ancestor so I dont want to do anything that conflicts with that. 

> Also, if you rebase-merge, you are creating new commits with the same diffs and messages as the PR's commits, but different SHAs. This is because rebase-merged commits can have different parent commits than the commits in the PR.

So by the sound of this, I shouldnt be doing a rebase-merge in GitHub. I also don't want ther to be excessive commits on master, so should I squash the commits locally before pushing upstream?

**Reply by airtower-luna** [reply]
> So by the sound of this, I shouldnt be doing a rebase-merge in GitHub. I also don't want ther to be excessive commits on master, so should I squash the commits locally before pushing upstream?

If you really want them squashed, yes. You'll need to then force-push to overwrite the previous state of the branch (potentially destructive, double-check before pushing). Then either do a regular merge, or fast-forward merge locally and push that (GitHub should recognize that and mark the PR as merged).

However, in my view squashing is the wrong choice in most cases. The rule of thumb is that a commit should be the smallest *logically complete* change (people might argue what is "logically complete"). Having small commits makes it easier to figure out where a bug was introduced: You can `git bisect` to find the commit, and if it's smaller it's easier to see what is wrong. Also if people write useful commit messages it's much easier to understand *why* something was done the way it is.

Stuff like WIP commits that *don't* compile or pass tests should definitely be fixed locally before merging, same if you want to clean up the history in other ways.

**Reply by plastikfan** [reply]
Thanks @airtower-luna, I mostly get what you're saying (I've never used ___git bisect___ so that's another rabbit hole I've gotta get into), but I also like to be practical. Using micro commits doesn't sound appealling especially for my own non-collaborative repo's and there are valid cases where a commit is neccessarily voluminous. In my personal opinion, placing arbitary limits on the size of a commit is making a rod for your back. For me what's more important are that the changes for a commit are cohesive and focused on the issue at hand, regardless of size. But I suppose, I could use your philosophy by limitng the scope of an issue, which may or may not be tricky, I'll have to see going forward.

**Comment by plastikfan** [comment]
After pruning with __git fetch --prune__ as suggested by @spenserblack the graph now re-assuringly looks like this:

![git-after-local-prune](https://user-images.githubusercontent.com/49785914/223199913-78b27456-3a50-4ed9-8f2d-8dc3a0c63b3c.jpg)

**Comment by plastikfan** [comment]
So after further digging around, I have come to the conclusion that there is no particular solution to the problem of stale local branches after the Pull Request has been merged back into the default branch. Ths is just a matter of ongoing maintenance, ie manually delete the local branch.

The biggest takeaway for me at least is @spenserblack's comment:

> Also, if you rebase-merge, you are creating new commits with the same diffs and messages as the PR's commits, but different SHAs. This is because rebase-merged commits can have different parent commits than the commits in the PR.

which helps me to understand why git can't associate the commits on the feat branch and the ones Pull Request merged, so thanks again @spenserblack.

See [git remove local merged branches](https://digitaldrummerj.me/git-remove-local-merged-branches/)

**Reply by spenserblack** [reply]
>so thanks again

No problem!

I did a little more digging (I just had to look up synonyms for "prune" :laughing:), and there are several projects called `git-trim` that can help with this maintenance. You might want to check out one of those or their competitors.

**Reply by plastikfan** [reply]
great, thanks @spenserblack

**Reply by plastikfan** [reply]
[git-trim](https://github.com/jasonmccreary/git-trim) lookings promising...
